llvm/lib/Target/AMDGPU/SIPreEmitPeephole.cpp

```diff
@@ -305,10 +310,53 @@ bool SIPreEmitPeephole::getBlockDestinations(
 }
 
 bool SIPreEmitPeephole::mustRetainExeczBranch(
-    const MachineBasicBlock &From, const MachineBasicBlock &To) const {
+    const MachineBasicBlock &Head, const MachineBasicBlock &From,
+    const MachineBasicBlock &To) const {
+
+  auto FromIt = find(Head.successors(), &From);
+  assert(FromIt != Head.succ_end());
+  BranchProbability ExecNZProb = Head.getSuccProbability(FromIt);
+
   unsigned NumInstr = 0;
-  const MachineFunction *MF = From.getParent();
 
+  unsigned long ExecNZBranchCost = 0;
+  unsigned long UnconditionalBranchCost = 0;
+  unsigned long N = 0;
+  unsigned long D = 0;
+  unsigned long ThenCyclesCost = 0;
+
+  std::function<bool(const MachineInstr &)> IsProfitable =
+      [&](const MachineInstr &MI) {
+        ++NumInstr;
+        if (NumInstr >= SkipThreshold)
+          return false;
+        // These instructions are potentially expensive even if EXEC = 0.
+        if (TII->isSMRD(MI) || TII->isVMEM(MI) || TII->isFLAT(MI) ||
+            TII->isDS(MI) || TII->isWaitcnt(MI.getOpcode()))
```

Comment:
I think we need to redo this logic, and is the harmful part for the use cases in question. The issue of the vmem instructions isn't expensive, it's waiting for them. In the case where they want to avoid the branch, there is a buffer load

Reply:
Just to be sure, what you want is to restrict the newly proposed cost-model (include the waitcnt restriction) ? 

Or to relax the "trivial" cost model (for example, remove the VMEM restriction) ?

I would be inclined to do both.

Reply:
Yes, both. As separate changes

Reply:
To remove the checks first from the old version: https://github.com/llvm/llvm-project/pull/109995

Reply:
I've added the isWaitcnt(MI.getOpcode()) in the isProfitable method.